src/Voron/Data/Fixed/FixedSizeTree.cs

```diff
@@ -120,52 +120,60 @@ private void ThrowScopeAlreadyOpen()
 
         }
 
-        public void RepurposeInstance(Slice treeName, bool clone)
+        public static bool TryRepurposeInstance(FixedSizeTree<TVal> tree, Slice treeName, bool clone)
         {
-            new DirectAddScope(this).Dispose();// verifying that we aren't holding a ptr out
+            new DirectAddScope(tree).Dispose();// verifying that we aren't holding a ptr out
 
+            // Setting the name of the tree has to happen before returning even if the return is null
+            // TODO: Make sure that is no longer the case by prohibiting the direct creation of trees.
             if (clone)
             {
-                if (_treeName.HasValue)
-                    _tx.Allocator.Release(ref _treeName.Content);
+                if (tree._treeName.HasValue)
+                    tree._tx.Allocator.Release(ref tree._treeName.Content);
 
-                _treeName = treeName.Clone(_tx.Allocator);
+                tree._treeName = treeName.Clone(tree._tx.Allocator);
             }
             else
             {
-                _treeName = treeName;
+                tree._treeName = treeName;
             }
 
-            var header = (FixedSizeTreeHeader.Embedded*)_parent.DirectRead(_treeName);
+            var header = (FixedSizeTreeHeader.Embedded*)tree._parent.DirectRead(treeName);
             if (header == null)
-                return;
+                return false;
```

Comment:
Yes, in fact I thought exactly as you the first time. Maybe  I should add more comments. It can happen that we try to repurpose the instance but the instance does not exist / or does not have any item in it. BUT, we should still have repurposed the instance for a new tree. This is one of the big inconsistencies I have to address but I am not addressing it yet.

Reply:
Any additional thoughts on this after spending more time on it?